## Task 
* A data scientist has been tasked by a US based Higher Education Institute with analyzing the provided dataset of Cyber Security roles to recommend two specializations for the study program, ensuring abundant opportunities for graduates upon program completion, along with a focus on achieving a good salary range.

In [1]:
import pandas as pd
salaries = pd.read_csv("salaries.csv")
salaries

work_year experience_level employment_type  \
0          2024               SE              FT   
1          2024               MI              CT   
2          2024               MI              CT   
3          2024               SE              FT   
4          2024               SE              FT   
...         ...              ...             ...   
6412       2020               MI              FT   
6413       2021               SE              FT   
6414       2021               SE              FT   
6415       2021               MI              FT   
6416       2021               MI              FT   

                              job_title  salary salary_currency  \
0             Cyber Security Consultant  175000             USD   
1     Vulnerability Management Engineer   50000             USD   
2     Vulnerability Management Engineer   50000             USD   
3                     Security Engineer  238050             USD   
4                     Security Engineer  146200             USD   
...                                 ...     ...             ...   
6412             Cyber Security Analyst  140000             AUD   
6413       Information Security Manager   60000             GBP   
6414       Penetration Testing Engineer  126000             USD   
6415       Information Security Analyst   42000             GBP   
6416        Threat Intelligence Analyst   66310             USD   

      salary_in_usd employee_residence  remote_ratio company_location  \
0            175000                 US           100               US   
1             50000                 CR             0               CR   
2             50000                 CR             0               CR   
3            238050                 US             0               US   
4            146200                 US             0               US   
...             ...                ...           ...              ...   
6412          96422                 AU            50               AU   
6413          82528                 GB            50               GB   
6414         126000                 US           100               US   
6415          57769                 GB           100               GB   
6416          66310                 US             0               US   

     company_size  
0               L  
1               M  
2               M  
3               M  
4               M  
...           ...  
6412            M  
6413            L  
6414            L  
6415            L  
6416            L  

[6417 rows x 11 columns]

In [2]:
salaries.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6417 entries, 0 to 6416
Data columns (total 11 columns):
 #   Column              Non-Null Count  Dtype 
---  ------              --------------  ----- 
 0   work_year           6417 non-null   int64 
 1   experience_level    6417 non-null   object
 2   employment_type     6417 non-null   object
 3   job_title           6417 non-null   object
 4   salary              6417 non-null   int64 
 5   salary_currency     6417 non-null   object
 6   salary_in_usd       6417 non-null   int64 
 7   employee_residence  6417 non-null   object
 8   remote_ratio        6417 non-null   int64 
 9   company_location    6417 non-null   object
 10  company_size        6417 non-null   object
dtypes: int64(4), object(7)
memory usage: 551.6+ KB


##### First Step
* Lets check if there is any value missing or in the file or any bad characters are overthere.

In [5]:
salaries.isnull().sum()

work_year             0
experience_level      0
employment_type       0
job_title             0
salary                0
salary_currency       0
salary_in_usd         0
employee_residence    0
remote_ratio          0
company_location      0
company_size          0
dtype: int64

* No empty value is overthere in all columns. 

#### Second step
* Lets create a fu7nction to find out if there is any special character or bad character is in file. 

In [11]:
def bad_cha(file_name):   
    bad_c = ["(",")"," ","c",".","'","s""-","_"]
    count = 0   
    for cl in file_name: # iterate over each column of the file
        for char in cl: # iterate over each cell of the column
            if char in bad_c:
                count +=1
    return count

In [12]:
file_name = 'salaries'
bad_cha(file_name)

0

In [13]:
# We can further verify if our function is working properly or not 

l = "60'50"
bad_cha(l)

1

* Its verified now that our file is clean. No need to work on cleaning the file.

#### Third step 
* After completing data cleaning on the file, the next step toward addressing our main objective is to identify which job title appears most frequently in the dataset. This will provide insight into the most common roles within the dataset, aiding in the decision-making process for specialization within the cybersecurity degree program.

In [14]:
job_title_counts = salaries['job_title'].value_counts()
print(job_title_counts)

job_title
Security Engineer                        2526
Security Analyst                          444
DevSecOps Engineer                        305
Security Architect                        222
Cyber Security Analyst                    169
                                         ... 
Cyber Security Lead                         1
Security Analytics Lead                     1
Business Information Security Officer       1
Systems Security Engineer                   1
Concierge Security Engineer                 1
Name: count, Length: 148, dtype: int64


* "Security Engineer" and "Security Analyst" are the most frequently occurring job titles in the dataset, it suggests a high demand for professionals in these roles within the field of cybersecurity. This information can inform decision-making regarding specialization within the cybersecurity degree program.

#### Fourth Step
* The next step is to analyze the experience level, remote job options, company size, and salary offered for the roles of "Security Engineer" and "Security Analyst."

In [15]:
# Filter out the "Security Engineer" record

specific_job_df = salaries[salaries['job_title'] == 'Security Engineer']
    
# Filter further to include only records where the employee residence is 'US'
specific_job_df_us = specific_job_df[specific_job_df['employee_residence'] == 'US']
print(len(specific_job_df_us))

2349


* Creating a function that takes two parameters: job_title and a DataFrame named "salaries". This function will generate a dictionary where each key represents an experience level, and its corresponding value is a list of salaries associated with that experience level. The function specifically focuses on jobs based in the US, ensuring that only salaries from US-based positions are included in the dictionary.

In [16]:
def salary(job_title, salaries):
  
    specific_job_df = salaries[salaries['job_title'] == job_title]
    
   
    specific_job_df_us = specific_job_df[specific_job_df['employee_residence'] == 'US']

    # Initialize dictionary to store salaries for each experience level
    salaries_by_experience = {
        "SE": [],
        "MI": [],
        "EN": [],
        "EX": []
    }

    # Iterate over each row index of the filtered DataFrame
    for index in specific_job_df_us.index:
        
        experience_level = specific_job_df_us.loc[index, 'experience_level']
        salary_in_usd = specific_job_df_us.loc[index, 'salary_in_usd']

        # Append the salary to the appropriate list based on experience level
        salaries_by_experience[experience_level].append(salary_in_usd)

    return salaries_by_experience


In [18]:
salary_Sec = salary('Security Engineer', salaries)
print(salary_Sec)

{'SE': [238050, 146200, 249800, 142700, 186000, 81800, 234000, 146000, 212000, 132000, 233000, 198000, 240000, 210000, 203550, 122400, 160000, 120000, 170000, 85000, 179000, 112000, 385000, 310000, 407000, 254000, 145000, 115000, 465000, 310000, 195330, 126990, 234000, 146000, 175160, 103035, 225000, 150000, 255200, 157800, 232200, 167200, 190000, 150000, 140000, 115000, 231000, 154000, 220000, 180000, 150000, 100000, 316000, 202000, 250000, 200000, 232000, 130000, 207000, 138000, 164749, 96025, 203044, 130180, 277000, 198000, 200000, 130000, 286700, 179200, 192500, 119000, 135000, 78000, 138000, 86000, 294000, 196000, 185000, 167000, 120000, 90000, 192500, 119000, 187000, 125000, 240000, 187000, 246000, 194000, 234000, 146000, 275000, 155500, 180000, 110000, 198000, 130000, 179000, 112000, 214500, 150000, 235400, 145500, 230000, 168000, 270000, 220500, 247500, 140000, 179000, 112000, 233500, 146000, 244860, 151400, 246000, 118000, 257900, 155200, 179000, 112000, 175100, 113000, 192500

In [43]:
# finding the length of SE role
len(salary_Sec['SE'])

1916

In [44]:
# finding the length of MI role
len(salary_Sec['MI'])

351

In [45]:
# finding the length of EN role
len(salary_Sec['EN'])

42

In [46]:
# finding the length of EX role
len(salary_Sec['EX'])

40

In [47]:
# Create a function to find out the average salary of a role

def ave_salary(salary_role, experience_level):
    
    total_sal_exp_lvl = sum(salary_role[experience_level])
    len_total_sal_exp_lvl = len(salary_role[experience_level])

    # Finding the average SE level salary

    ave_sal_exp_lvl = total_sal_exp_lvl//len_total_sal_exp_lvl
    print(ave_sal_exp_lvl)

In [48]:
ave_salary(salary_Sec,'SE')

177996


In [49]:
ave_salary(salary_Sec, 'MI')

146808


In [50]:
ave_salary(salary_Sec, 'EN')

119528


In [51]:
ave_salary(salary_Sec, 'EX')

215448


In [52]:
salary_Ana = salary('Security Analyst', salaries)
print(salary_Ana)

{'SE': [138000, 86000, 179000, 112000, 161500, 89700, 153900, 85500, 85000, 75000, 175000, 125000, 115920, 84000, 200000, 96000, 168715, 99230, 246000, 118000, 145000, 125000, 119000, 52100, 168715, 99230, 179000, 112000, 138000, 86000, 193866, 114039, 190000, 150000, 133000, 104910, 134000, 90000, 145000, 120000, 178250, 88400, 212568, 132855, 133000, 58300, 166000, 73000, 97000, 65000, 175000, 157000, 134000, 90000, 130000, 110000, 102793, 71281, 115000, 70000, 130000, 95000, 175000, 125000, 285000, 95000, 132170, 105740, 125000, 95000, 160000, 120000, 152570, 89300, 115500, 71400, 130000, 107000, 165000, 140250, 222000, 115920, 179000, 140000, 95000, 75000, 123000, 96000, 112000, 88000, 179000, 130000, 179000, 130000, 145000, 100000, 90600, 72500, 198900, 75000, 90600, 72500, 211750, 192500, 95000, 75000, 179000, 140000, 150300, 59800, 100000, 90000, 100000, 90000, 80000, 64000, 120350, 74400, 110400, 68000, 120000, 90000, 135000, 110000, 185900, 129300, 154300, 90000, 155500, 12750

In [53]:
# finding the length of SE role in Security Analyst
len(salary_Ana['SE'])

196

In [54]:
# finding the length of SE role in Security Analyst
len(salary_Ana['MI'])

115

In [55]:
# finding the length of SE role in Security Analyst
len(salary_Ana['EN'])

95

In [56]:
# finding the length of SE role in Security Analyst
len(salary_Ana['EX'])

4

In [58]:
# average salary of secuirty analyst as a role of SE
ave_salary(salary_Ana, 'SE')

128534


In [59]:
# average salary of secuirty analyst as a role of MI
ave_salary(salary_Ana, 'MI')

112448


In [60]:
# average salary of secuirty analyst as a role of EX
ave_salary(salary_Ana, 'EX')

133600


In [61]:
# average salary of secuirty analyst as a role of EN
ave_salary(salary_Ana, 'EN')

106688


#### Result of Salaries as a Security Engineer within USA
* The average salary of SE role is 177996 USD
* The average salary of MI role is 146808 USD 
* The average salary of EN role is 119528 USD
* The average salary of EX role is 215448 USD

* Majorly positions are occupied as SE role (1916)
* Total positions as Security Engineer are (2349)

#### Result of Salaries as a Security Analyst within USA
* The average salary of SE role is 128534 USD
* The average salary of MI role is 112448 USD
* The average salary of EN role is 106688 USD
* The average salary of EX role is 133600 USD
* Majorly positions are occupied as SE role (196)
* Total positions as Security Analyst are (410)

In [62]:
# Creating a function for finding out the size of the companies for both role inside the US. 

def company_size(job_title, salaries):

    specific_job_df = salaries[salaries['job_title'] == job_title]
    specific_job_df_us = specific_job_df[specific_job_df['employee_residence'] == 'US']
    # Access the 'company_size' information for the specific job title
    company_size = specific_job_df_us['company_size']

    # Initialize a dictionary to store company sizes for each group
    company_sizes_dict = {
        'S': [],
        'M': [],
        'L': []
    }

    # Iterate over each company size for the specific job title
    for size in company_size:
        if size == "S":
            company_sizes_dict['S'].append(size)
        elif size == "M":
            company_sizes_dict['M'].append(size)
        elif size == "L":
            company_sizes_dict['L'].append(size)
    return company_sizes_dict

#### Finding out the length of all size of companies for Security Engineer

In [73]:
size = company_size('Security Engineer', salaries)
print("Number of small companies for Security Engineer: ",len(size['S']))

Number of small companies for Security Engineer:  6


In [77]:
print("Number of medium companies for Security Engineer: ",len(size['M']))

Number of medium companies for Security Engineer:  2096


In [78]:
print("Number of large companies for Security Engineer: ",len(size['L']))

Number of large companies for Security Engineer:  247


#### Finding out the length of all size of companies for Security Analyst


In [79]:
size = company_size('Security Analyst', salaries)
print("Number of small companies for Security Analyst: ",len(size['S']))
print("Number of medium companies for Security Analyst: ",len(size['M']))
print("Number of large companies for Security Analyst: ",len(size['L']))

Number of small companies for Security Analyst:  5
Number of medium companies for Security Analyst:  373
Number of large companies for Security Analyst:  32


### Findings from companies sizes

* The number of medium and large size is more comparatively to small size companies which are hiring Security Engineers and Security Analysts which shows that starting specializations in these fields is quite important and will be worthful for the future of the students as well of the institute. 

#### Remote Jobs 
* There is one more factor which can be useful for the students and they can opt these specializations. Its remote job. 


In [84]:
# Create a function to find out the remote jobs choice 

def count_remote_choices(job_title, salaries):
   
    #
    specific_job_df = salaries[salaries['job_title'] == job_title]
    specific_job_df_us = specific_job_df[specific_job_df['employee_residence'] == 'US']
    
    # Access the 'remote_job_choice' information for the specific job title with US
    remote_ratio_specific_job = specific_job_df_us['remote_ratio']

    # Initialize counts
    count_100 = 0
    count_50 = 0
    count_0 = 0

    # Iterate over each remote job choice for the specific job title
    for remote_choice in remote_ratio_specific_job:
        if remote_choice == 100:
            count_100 += 1
        elif remote_choice == 50:
            count_50 += 1
        else:
            count_0 += 1

    # Create a dictionary to store the counts
    counts_dict = {
        '100% remote': count_100,
        '50% remote': count_50,
        '0% remote': count_0
    }

    return counts_dict


counts = count_remote_choices('Security Engineer', salaries)
print("Counts for Security Engineer jobs:", counts)


Counts for Security Engineer jobs: {'100% remote': 744, '50% remote': 8, '0% remote': 1597}


In [88]:
# total jobs for "Security Engineer" role is already calculated previously 

percentage_remoteJob_SE = (744/2349)*100 
percentage_remoteJob_SE

31.673052362707537

In [86]:
# Finding out the remote jobs for Security Analyst

counts_SA = count_remote_choices('Security Analyst', salaries)
print("Counts for Security Engineer jobs:", counts_SA)

Counts for Security Engineer jobs: {'100% remote': 143, '50% remote': 2, '0% remote': 265}


In [89]:
# len(salary_Ana['SE']) is already calculated previously 
percentage_remoteJob_SA = (143/410)*100
percentage_remoteJob_SA

34.8780487804878

#### Findings from Remote jobs:
* 31% of the total positions of Security Engineer in US are remote jobs.
* 34% of the total positions of Security Analyst in US are remote.

In [90]:
### Creating a function yo check the number of positions of a role in different years

def count_jobs_by_year(job_title, work_year, salaries):
   
    specific_job_df = salaries[salaries['job_title'] == job_title]
    
    # Filter further to include only records where the employee residence is 'US' and work year matches
    specific_job_df_us = specific_job_df[(specific_job_df['employee_residence'] == 'US') & 
                                         (specific_job_df['work_year'] == work_year)]
    
   
    work_year_specific_job = specific_job_df_us['work_year']

    # Initialize count for jobs in the specified year
    count_jobs = len(work_year_specific_job)
    
    return count_jobs


### Finding jobs from 2020-2024 in Security Engineering 

In [91]:
jobs_SecEng_2020 = count_jobs_by_year('Security Engineer', 2020, salaries)
jobs_SecEng_2020

10

In [93]:
jobs_SecEng_2021 = count_jobs_by_year('Security Engineer', 2021, salaries)
jobs_SecEng_2021

9

In [94]:
jobs_SecEng_2022 = count_jobs_by_year('Security Engineer', 2022, salaries)
jobs_SecEng_2022

297

In [95]:
jobs_SecEng_2023 = count_jobs_by_year('Security Engineer', 2023, salaries)
jobs_SecEng_2023

1606

In [96]:
jobs_SecEng_2024 = count_jobs_by_year('Security Engineer', 2024, salaries)
jobs_SecEng_2024

427

### Finding jobs from 2020-2024 in Security Analyst

In [100]:
jobs_SecEng_2020 = count_jobs_by_year('Security Analyst', 2020, salaries)
print(jobs_SecEng_2020)
jobs_SecEng_2021 = count_jobs_by_year('Security Analyst', 2021, salaries)
print(jobs_SecEng_2021)
jobs_SecEng_2022 = count_jobs_by_year('Security Analyst', 2022, salaries)
print(jobs_SecEng_2022)
jobs_SecEng_2023 = count_jobs_by_year('Security Analyst', 2023, salaries)
print(jobs_SecEng_2023)
jobs_SecEng_2024 = count_jobs_by_year('Security Analyst', 2024, salaries)
print(jobs_SecEng_2024)

2
14
69
221
104


### Job growth in Security Engineer and Analyst in US
* Accoding to the data statistics, job growth in the US in the field of Security Engineer is increasing.
* In 2023 compare to 2022, almost 6 times job growth is observed in the field of Security Engineer. 
* In 2024 (Though its just beginning of 2024) we have noticed 2 times more positions in the field of security engineer

* Similarly compare to 2022, almost 3 times more positions are introduced in the field of Security Analyst.
* A significant number of new positions in the field of Security Analyst is observed in 2024 till now as well.